bindings/rust/s2n-tls/src/renegotiate.rs

```diff
@@ -0,0 +1,677 @@
+        let context = self.config.context_mut();
+        context.renegotiate = Some(handler);
+        unsafe {
+            s2n_config_set_renegotiate_request_cb(
+                self.as_mut_ptr(),
+                Some(renegotiate_cb),
+                std::ptr::null_mut(),
+            )
+            .into_result()?;
+        }
+        Ok(self)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{
+        callbacks::{
+            ConnectionFuture, ConnectionFutureResult, PrivateKeyCallback, PrivateKeyOperation,
+        },
+        config::ConnectionInitializer,
+        testing::{CertKeyPair, InsecureAcceptAllCertificatesHandler, TestPair, TestPairIO},
+    };
+    use foreign_types::ForeignTypeRef;
+    use futures_test::task::new_count_waker;
+    use openssl::ssl::{
+        ErrorCode, NameType, Ssl, SslContext, SslFiletype, SslMethod, SslStream, SslVerifyMode,
+        SslVersion,
+    };
+    use std::{
+        error::Error,
+        io::{Read, Write},
+        pin::Pin,
+        task::Poll::{Pending, Ready},
+    };
+
+    // Currently renegotiation is not available from the openssl-sys bindings
+    extern "C" {
+        fn SSL_renegotiate(s: *mut openssl_sys::SSL) -> libc::size_t;
+        fn SSL_renegotiate_pending(s: *mut openssl_sys::SSL) -> libc::size_t;
+    }
+
+    // std::task::ready is unstable
+    fn unwrap_poll<T>(
+        poll: Poll<Result<T, crate::error::Error>>,
+    ) -> Result<(), crate::error::Error> {
+        if let Ready(value) = poll {
+            return value.map(|_| ());
+        }
+        panic!("Poll not Ready");
+    }
+
+    #[derive(Debug)]
+    struct ServerTestStream(TestPairIO);
+
+    // For server testing purposes, we read from the client output stream
+    impl Read for ServerTestStream {
+        fn read(&mut self, buf: &mut [u8]) -> Result<usize, std::io::Error> {
+            let result = self.0.client_tx_stream.borrow_mut().read(buf);
+            if let Ok(0) = result {
+                // Treat no data as blocking instead of EOF
+                Err(std::io::Error::new(
+                    std::io::ErrorKind::WouldBlock,
+                    "blocking",
+                ))
+            } else {
+                result
+            }
+        }
+    }
+
+    // For server testing purposes, we write to the server output stream
+    impl Write for ServerTestStream {
+        fn write(&mut self, buf: &[u8]) -> Result<usize, std::io::Error> {
+            self.0.server_tx_stream.borrow_mut().write(buf)
+        }
+
+        fn flush(&mut self) -> Result<(), std::io::Error> {
+            self.0.server_tx_stream.borrow_mut().flush()
+        }
+    }
+
+    // s2n-tls doesn't support sending client hello requests.
+    // This makes it impossible to test renegotiation without direct access to
+    // s2n-tls internals like the methods for sending arbitrary records.
+    // Instead, we need to use openssl as our server.
+    struct RenegotiateTestPair {
```

Comment:
I think the IO on this is a bit lopsided, in that both `client` and `server` have references to `TestPairIO`, but `server` somewhat arbitrarily owns it.

For the purposes of a test fixture I don't think that's a bit deal, but I think we should at least include a comment about it. "For convenience with OpenSSL's API the TestPairIO is owns by SSLStream", etc.